In [ ]:
### elmeleti preziben elmondani: import modul
### bemutatni hogy milyen formaban van az adata: cel 3 rendes adattabla

## GOAL: have separate lists for the three data filenames

# Example

### Listing files in directory with os modul

In [1]:
import os

In [59]:
os.getcwd()

'/home/esztersomos/Documents/CEU/winter_2019/load_data'

In [83]:
os.chdir('/home/esztersomos/Documents/CEU/winter_2019/load_data/example_data_structure_folder/')

In [84]:
os.getcwd()

'/home/esztersomos/Documents/CEU/winter_2019/load_data/example_data_structure_folder'

In [5]:
filenames = os.listdir()

### Lists

In [6]:
filenames

['b_1.csv', 'a_2.csv', 'b_3.csv', 'b_2.csv', 'a_1.csv', 'a_3.csv']

In [10]:
type(filenames)

list

In [8]:
len(filenames)

6

In [22]:
a_data_list = []
b_data_list = []

In [23]:
type(a_data_list), len(a_data_list)

(list, 0)

### Slicing lists and strings

In [16]:
first_filename = filenames[0]

In [17]:
first_filename

'b_1.csv'

In [18]:
type(first_filename)

str

In [19]:
first_char = first_filename[0]
first_char

'b'

### Looping lists

In [13]:
for f in filenames:
    print(f)

b_1.csv
a_2.csv
b_3.csv
b_2.csv
a_1.csv
a_3.csv


In [20]:
for f in filenames:
    print(f[0])

b
a
b
b
a
a


### If statement

In [24]:
first_filename

'b_1.csv'

In [25]:
first_filename[0] == 'b'

True

In [26]:
if first_filename[0] == 'b':
    print('This filename starts with b')

This filename starts with b


In [28]:
second_filename = filenames[1]
second_filename

'a_2.csv'

In [29]:
if second_filename[0] == 'b':
    print('This filename starts with b')

### Using if statement inside a loop

In [30]:
for f in filenames:
    print(f)
    if f[0] == 'a':
        print('This filename starts with a')
    if f[0] == 'b':
        print('This filename starts with b')

b_1.csv
This filename starts with b
a_2.csv
This filename starts with a
b_3.csv
This filename starts with b
b_2.csv
This filename starts with b
a_1.csv
This filename starts with a
a_3.csv
This filename starts with a


### Adding to list

In [36]:
mylist = []

In [37]:
mylist

[]

In [38]:
mylist.append('new_element')

In [39]:
mylist

['new_element']

In [40]:
myvariable = 'newer_element'
mylist.append(myvariable)

In [41]:
mylist

['new_element', 'newer_element']

### Adding elements to list inside a loop based on the result of if statement

In [47]:
a_data_list = []
b_data_list = []

In [48]:
a_data_list

[]

In [49]:
b_data_list

[]

In [50]:
for f in filenames:
    print(f)
    if f[0] == 'a':
        print('This filename starts with a so it will be appended to a_data_list')
        a_data_list.append(f)
    if f[0] == 'b':
        print('This filename starts with b so it will be appended to b_data_list')
        b_data_list.append(f)
    print('a_data_list at this point is:')
    print(a_data_list)
    print('b_data_list at this point is:')
    print(b_data_list)
    print()

b_1.csv
This filename starts with b so it will be appended to b_data_list
a_data_list at this point is:
[]
b_data_list at this point is:
['b_1.csv']

a_2.csv
This filename starts with a so it will be appended to a_data_list
a_data_list at this point is:
['a_2.csv']
b_data_list at this point is:
['b_1.csv']

b_3.csv
This filename starts with b so it will be appended to b_data_list
a_data_list at this point is:
['a_2.csv']
b_data_list at this point is:
['b_1.csv', 'b_3.csv']

b_2.csv
This filename starts with b so it will be appended to b_data_list
a_data_list at this point is:
['a_2.csv']
b_data_list at this point is:
['b_1.csv', 'b_3.csv', 'b_2.csv']

a_1.csv
This filename starts with a so it will be appended to a_data_list
a_data_list at this point is:
['a_2.csv', 'a_1.csv']
b_data_list at this point is:
['b_1.csv', 'b_3.csv', 'b_2.csv']

a_3.csv
This filename starts with a so it will be appended to a_data_list
a_data_list at this point is:
['a_2.csv', 'a_1.csv', 'a_3.csv']
b_data_lis

### Sorting lists

In [125]:
a_data_list

['a_2.csv', 'a_1.csv', 'a_3.csv']

In [127]:
sorted_a_data_list = sorted(a_data_list)

In [128]:
sorted_a_data_list

['a_1.csv', 'a_2.csv', 'a_3.csv']

### pandas dataframe

In [9]:
import pandas as pd

In [141]:
first_a_data_filename = sorted_a_data_list[0]
first_a_data_filename

'a_1.csv'

In [142]:
first_a_df = pd.read_csv(first_a_data_filename)

In [143]:
type(first_a_df)

pandas.core.frame.DataFrame

In [144]:
first_a_df

a aa
0  1  a
1  2  b

In [145]:
len(first_a_df)

2

In [146]:
second_a_data_filename = sorted_a_data_list[1]
second_a_data_filename

'a_2.csv'

In [147]:
second_a_df = pd.read_csv(second_a_data_filename)

In [148]:
second_a_df

a aa
0  3  c
1  4  d

### Concatenating dataframes

In [149]:
a_dataframes = [first_a_df, second_a_df]

In [150]:
type(a_dataframes)

list

In [151]:
a_dataframes[0]

a aa
0  1  a
1  2  b

In [152]:
type(a_dataframes[0])

pandas.core.frame.DataFrame

In [153]:
concatenated_a_df = pd.concat(a_dataframes)

In [154]:
concatenated_a_df

a aa
0  1  a
1  2  b
0  3  c
1  4  d

In [155]:
len(concatenated_a_df)

4

In [156]:
a_dataframes = []
for a_filename in sorted_a_data_list:
    a_df = pd.read_csv(a_filename)
    a_dataframes.append(a_df)

In [157]:
len(a_dataframes)

3

In [158]:
type(a_dataframes[0])

pandas.core.frame.DataFrame

In [159]:
a_dataframes[0]

a aa
0  1  a
1  2  b

In [160]:
concatenated_a_df = pd.concat(a_dataframes)

In [161]:
concatenated_a_df

a aa
0  1  a
1  2  b
0  3  c
1  4  d
0  5  e
1  6  f

In [162]:
sorted_b_data_list = sorted(b_data_list)
b_dataframes = []
for b_filename in b_data_list:
    b_df = pd.read_csv(b_filename)
    b_dataframes.append(b_df)
concatenated_b_df = pd.concat(b_dataframes)

In [163]:
concatenated_b_df

b  bb
0  10  aa
1  20  bb
0  50  ee
1  60  ff
0  30  cc
1  40  dd

### Saving dataframes as csv

In [200]:
concatenated_a_df.to_csv('a.csv', index=False)

# Excercise

### 1 - exercise
Import the os modul and look at the current working directory. If this is not the load_data directory, use the os.chdir() command to change the working directory! 

In [7]:
### Your code here
import os
import pandas as pd
os.getcwd()

'/home/esztersomos/Documents/CEU/winter_2019/load_data'

### 1 - check yourself
The output of this cell should end with "/load_data"

In [15]:
os.getcwd()

'/home/esztersomos/Documents/CEU/winter_2019/load_data'

### 2 - exercise
Create a variable called filenames that contains the files and folders in the load_data library. (Use the os.listdir() command for this) <br>
Create three empty lists with the names: trip_file_list, station_file_list, weather_file_list

In [13]:
os.chdir('/home/esztersomos/Documents/CEU/winter_2019/load_data/')

In [2]:
### Your code here
filenames = os.listdir()
trip_file_list = []
station_file_list = []
weather_file_list = []

### 2 - check yourself
The length of the filenames list should be 169 (or 168 if you haven't saved this notebook yet). <br>
The length of the other lists should be 0

In [3]:
print('Length of filenames list is {}'.format(len(filenames)))
print('Length of trip_file_list is {}'.format(len(trip_file_list)))
print('Length of station_file_list list is {}'.format(len(station_file_list)))
print('Length of weather_file_list is {}'.format(len(weather_file_list)))

Length of filenames list is 169
Length of trip_file_list is 0
Length of station_file_list list is 0
Length of weather_file_list is 0


### 3 - exercise
Write a loop on the filenames list. Inside the loop check if the filenames starts with 't'. If it does, append it to the trip_file_list. Do similarly for the station_file_list and for the weather_file_list

Bonus exercise: Instead of using 3 if statements, try using 1 if-elif-else statement.

In [4]:
### Your code here
for f in filenames:
    if f[0] == 't':
        trip_file_list.append(f)
    if f[0] == 's':
        station_file_list.append(f)
    if f[0] == 'w':
        weather_file_list.append(f)
print(len(trip_file_list), len(station_file_list), len(weather_file_list))


145 2 19


### 3 - check yourself
The length of the trip_file_list should be 145 <br>
The length of the station_file_list should be 2 <br>
The length of the weather_file_list should be 19 

In [5]:
print('Length of trip_file_list is {}'.format(len(trip_file_list)))
print('Length of station_file_list list is {}'.format(len(station_file_list)))
print('Length of weather_file_list is {}'.format(len(weather_file_list)))

Length of trip_file_list is 145
Length of station_file_list list is 2
Length of weather_file_list is 19


In [ ]:
### legyen ide olyan hogy ellenorizni kell hogy mindegyik megvan-e? es akkor split, int(), range()

### 4 - exercise
Sort the trip, station and weather file lists. Make 3 dataframes by reading in the csv named as the first element from each lists. <br><br>
At the end should have a dataframe called first_trip_df containing the trip_1.csv data, <br>
a dataframe called first_station_df containing the station_1.csv data and <br>
a dataframe called first_weather_df containing the weather_1.csv data.

In [8]:
### Your code here
os.chdir('/home/esztersomos/Documents/CEU/winter_2019/load_data/')
first_trip_df = pd.read_csv(sorted(trip_file_list)[0])
first_station_df = pd.read_csv(sorted(station_file_list)[0])
first_weather_df = pd.read_csv(sorted(weather_file_list)[0])

### 4 - check yourself
The Trip ID in the second row of the first_trip_df should be 4607 <br>
The station_id in the second row of the first_trip_df should be 3 <br>
The date in the second row of the first_trip_df should be 8/30/2013 <br>

In [9]:
print('This is the first 5 rows from the first_trip_df dataframe')
display(first_trip_df.head())
print('\nThis is the first 5 rows from the first_station_df dataframe')
display(first_station_df.head())
print('\nThis is the first 5 rows from the first_weather_df dataframe')
display(first_weather_df.head())

This is the first 5 rows from the first_trip_df dataframe


Trip ID       Start Date             Start Station  Start Terminal  \
0     4576  8/29/2013 14:13  South Van Ness at Market              66   
1     4607  8/29/2013 14:42        San Jose City Hall              10   
2     4130  8/29/2013 10:16   Mountain View City Hall              27   
3     4251  8/29/2013 11:29        San Jose City Hall              10   
4     4299  8/29/2013 12:02  South Van Ness at Market              66   

          End Date               End Station  End Terminal  Bike #  \
0  8/29/2013 14:14  South Van Ness at Market            66     520   
1  8/29/2013 14:43        San Jose City Hall            10     661   
2  8/29/2013 10:17   Mountain View City Hall            27      48   
3  8/29/2013 11:30        San Jose City Hall            10      26   
4  8/29/2013 12:04            Market at 10th            67     319   

  Subscription Type  Zip Code  
0        Subscriber   94127.0  
1        Subscriber   95138.0  
2        Subscriber   97214.0  
3        Subscriber   95060.0  
4        Subscriber   94103.0


This is the first 5 rows from the first_station_df dataframe


station_id                               name        lat        long  \
0           2  San Jose Diridon Caltrain Station  37.329732 -121.901782   
1           3              San Jose Civic Center  37.330698 -121.888979   
2           4             Santa Clara at Almaden  37.333988 -121.894902   
3           5                   Adobe on Almaden  37.331415 -121.893200   
4           6                   San Pedro Square  37.336721 -121.894074   

   dockcount  landmark installation  
0         27  San Jose     8/6/2013  
1         15  San Jose     8/5/2013  
2         11  San Jose     8/6/2013  
3         19  San Jose     8/5/2013  
4         15  San Jose     8/7/2013


This is the first 5 rows from the first_weather_df dataframe


Date  Max_Temperature_F  Mean_Temperature_F  Min_TemperatureF  \
0  8/29/2013               74.0                68.0              61.0   
1  8/30/2013               78.0                69.0              60.0   
2  8/31/2013               71.0                64.0              57.0   
3   9/1/2013               74.0                66.0              58.0   
4   9/2/2013               75.0                69.0              62.0   

   Max_Dew_Point_F  MeanDew_Point_F  Min_Dewpoint_F  Max_Humidity  \
0               61               58              56            93   
1               61               58              56            90   
2               57               56              54            93   
3               60               56              53            87   
4               61               60              58            93   

   Mean_Humidity   Min_Humidity   ...    Mean_Visibility_Miles   \
0              75             57  ...                        10   
1              70             50  ...                        10   
2              75             57  ...                        10   
3              68             49  ...                        10   
4              77             61  ...                        10   

   Min_Visibility_Miles   Max_Wind_Speed_MPH   Mean_Wind_Speed_MPH   \
0                     10                   23                    11   
1                      7                   29                    13   
2                     10                   26                    15   
3                     10                   25                    13   
4                      6                   23                    12   

   Max_Gust_Speed_MPH  Precipitation_In   Cloud_Cover   Events  \
0                28.0                  0             4     NaN   
1                35.0                  0             2     NaN   
2                31.0                  0             4     NaN   
3                29.0                  0             4     NaN   
4                30.0                  0             6     NaN   

   Wind_Dir_Degrees    zip  
0               286  94107  
1               291  94107  
2               284  94107  
3               284  94107  
4               277  94107  

[5 rows x 24 columns]

### 5 - exercise
Create an empty list called trip_dataframes <br>
Loop through the filenames in the sorted trip file list. Inside the loop read in the corresponding csv into a dataframe. Append this dataframe to the trip_dataframes. <br>
After the loop finished concatenate the dataframes in the trip_dataframes list into a dataframe called trip. <br><br>
Similarly create a station and a weather dataframe.

In [11]:
### Your code here
trip_dataframes = []
for f in sorted(trip_file_list):
    trip_dataframes.append(pd.read_csv(f))
trip = pd.concat(trip_dataframes)

station_dataframes = []
for f in sorted(station_file_list):
    station_dataframes.append(pd.read_csv(f))
station = pd.concat(station_dataframes)

weather_dataframes = []
for f in sorted(weather_file_list):
    weather_dataframes.append(pd.read_csv(f))
weather = pd.concat(weather_dataframes)

### 5 - check yourself
The length of the trip dataframe should be 144115 <br>
The length of the station_file_list should be 69 <br>
The length of the weather_file_list should be 928

In [23]:
print('Length of trip dataframe is {}'.format(len(trip)))
print('Length of station dataframe list is {}'.format(len(station)))
print('Length of weather dataframe is {}'.format(len(weather)))

Length of trip dataframe is 144115
Length of station dataframe list is 69
Length of weather dataframe is 928


### 6 - exercise
Write the trip, the station and the weather dataframes in csv files called trip.csv, station.csv and weather.csv. <br>
Use the index=False argument

In [16]:
os.chdir('/home/esztersomos/Documents/CEU/winter_2019/check_data/')

In [17]:
### Your code here
trip.to_csv('trip.csv', index=False)
station.to_csv('station.csv', index=False)
weather.to_csv('weather.csv', index=False)


### 6 - check yourself
Run the code cell below to check this task

In [26]:
if 'trip.csv' in os.listdir():
    print('trip.csv was successfully saved')
else:
    print('trip.csv was NOT successfully saved')
if 'station.csv' in os.listdir():
    print('station.csv was successfully saved')
else:
    print('station.csv was NOT successfully saved')
if 'weather.csv' in os.listdir():
    print('weather.csv was successfully saved')
else:
    print('weather.csv was NOT successfully saved')

trip.csv was successfully saved
station.csv was successfully saved
weather.csv was successfully saved
